root: Restructure broker / cache / channel / result configuration

PKizzle opened the pull request:
This is part of https://github.com/goauthentik/authentik/pull/5395.

It restructures the configuration to separate Redis configuration clearly from cache and broker settings.

Furthermore, it enables custom configuration of the broker with its respective transport options, cache, channel and result backend. With this change i.e. RabbitMQ or Amazon SQS may be used as a broker.
Broker transport options need to be Base64 encoded and are ignored if not formatted correctly.

Deprecation logic has been added that automatically converts the old configuration names into the new ones and outputs a log message to encourage changing the configuration accordingly.

PKizzle committed:
Initial commit

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

PKizzle committed:
Remove any remaining mentions of Redis URL

This is handled in https://github.com/goauthentik/authentik/pull/5395

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

# Conflicts:
#	authentik/lib/config.py

PKizzle committed:
Allow setting broker transport options

This enables usage of other brokers that require additional settings

PKizzle committed:
Remove remaining reference to Redis URL

This functionality is not part of this PR

PKizzle committed:
Reset default TLS requirements to none

PKizzle committed:
Fix linter errors

PKizzle committed:
Move dict from base64 encoded json to config.py

Additionally add tests

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

PKizzle commented:
@BeryJu This PR is ready to be reviewed in preparation for https://github.com/goauthentik/authentik/pull/7118.

BeryJu commented on authentik/root/settings.py:
```diff
@@ -193,9 +192,9 @@
-        "BACKEND": "django_redis.cache.RedisCache",
-        "LOCATION": f"{_redis_url}/{CONFIG.get('redis.db')}",
-        "TIMEOUT": CONFIG.get_int("redis.cache_timeout", 300),
+        "BACKEND": CONFIG.get("cache.backend", "django_redis.cache.RedisCache"),
```
I'm not sure we want to expose this as a config option as we only have the one backend?

PKizzle replied:
This gives the user the freedom to choose and use whatever backend might come in handy.
Authentik can still only offer support for Redis. For example one might want to use a database as cache instead.

BeryJu replied:
I'm tempted to point people to https://goauthentik.io/docs/installation/configuration#custom-python-settings for this instead of having that in the native config options (mainly going by only exposing config options where we support all possible options)

BeryJu commented on authentik/root/settings.py:
```diff
@@ -253,9 +252,9 @@
 CHANNEL_LAYERS = {
     "default": {
-        "BACKEND": "channels_redis.core.RedisChannelLayer",
+        "BACKEND": CONFIG.get("channel.backend", "channels_redis.core.RedisChannelLayer"),
```
Same as above

PKizzle commented on authentik/root/settings.py:
```diff
@@ -330,6 +331,23 @@
+            _broker_transport_options_str.strip().lstrip("{").rstrip("}")
+        )
+        _broker_transport_options_str = "{" + _broker_transport_options_str + "}"
+        _broker_transport_options.update(ast.literal_eval(_broker_transport_options_str))
```
I was not sure whether all possible `broker_transport_options` adhere to the JSON format.

PKizzle replied:
I have changed it to `json.loads`.

PKizzle committed:
Replace ast.literal_eval with json.loads

PKizzle committed:
Use default channel and cache backend configuration

If more customization is desired users shall look at goauthentik.io/docs/installation/configuration#custom-python-settings

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

PKizzle commented:
@BeryJu Could you take a look whether there are any more changes I should make?

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

# Conflicts:
#	authentik/root/settings.py

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

BeryJu commented on authentik/lib/config.py:
```diff
@@ -22,6 +24,24 @@
 
+REDIS_ENV_KEYS = [
+    f"{ENV_PREFIX}_REDIS__HOST",
+    f"{ENV_PREFIX}_REDIS__PORT" f"{ENV_PREFIX}_REDIS__DB",
```
```suggestion
    f"{ENV_PREFIX}_REDIS__PORT",
    f"{ENV_PREFIX}_REDIS__DB",
```

PKizzle replied:
Fixed.

BeryJu commented on authentik/lib/config.py:
```diff
@@ -180,6 +229,21 @@ def update_from_file(self, path: Path):
+        self.__config.update(update)
+
+    @staticmethod
+    def _set_value_for_key_path(outer, path, value, sep="."):
```
I think this is due to the rebase, but `set_path_in_dict` should do the same thing as this

PKizzle replied:
I have remove the duplicate.

BeryJu commented on website/docs/installation/configuration.mdx:
```diff
@@ -71,16 +71,38 @@ To check if your config has been applied correctly, you can run the following co
+    `AUTHENTIK_CACHE__TIMEOUT_REPUTATION` only applies to the cache expiry, see [`AUTHENTIK_REPUTATION__EXPIRY`](#authentik_reputation__expiry) to control how long reputation is persisted for.
+    :::
+
+## Channel Settings
```
We should probably explain for the docs what "Channel" means

```suggestion
## Websocket Settings
```

PKizzle replied:
I would like to keep the reference to Django Channel Layers as it makes it much easier to lookup should any questions arise.

BeryJu commented on authentik/lib/config.py:
```diff
@@ -188,19 +252,12 @@ def update_from_env(self):
-                current_obj = current_obj[dot_part]
             # Check if the value is json, and try to load it
             try:
                 value = loads(value)
```
Could we try to parse and load b64 here instead of explicitly having to call `get_dict_from_b64_json`? Maybe not such a good idea since some values might be b64 encoded that shouldn't be decoded?

PKizzle replied:
Furthermore, b64 encoded values can then only be set in env vars and not in the yaml files.

BeryJu reviewed (changes requested):
Only some small nits/questions, otherwise LGTM

rissson commented on authentik/lib/config.py:
```diff
@@ -113,6 +137,31 @@ def __init__(self, **kwargs):
+
+        for deprecation, replacement in DEPRECATIONS.items():
+            if self.get(deprecation, default=UNSET) is not UNSET:
+                self.log(
```
We may also want to send notifications to the admins here.

PKizzle replied:
How does one send notifications to admins?

PKizzle replied:
Using `Notifications.objects.create` ?

BeryJu replied:
it would be with `Event.new` with the action set to `configuration_error`, but the config loader purposefully doesn't require any database model to be loaded so I'd prefer to keep it that way (maybe we could have a dynamic import and handle it failing?)

PKizzle committed:
Send config deprecation notification to all superusers

PKizzle committed:
Remove duplicate method

PKizzle committed:
Merge branch 'main' into feature/reorganize-redis-config

PKizzle committed:
Add configuration explanation

For channel layer settings

PKizzle committed:
Use Event for deprecation warning

PKizzle committed:
Fix remove duplicated method

PKizzle committed:
Add missing comma

BeryJu commented on authentik/lib/config.py:
```diff
@@ -113,6 +138,40 @@ def __init__(self, **kwargs):
+                try:
+                    from authentik.events.models import Event, EventAction
+
+                    Event.new(EventAction.CONFIGURATION_ERROR, message=message)
```
```suggestion
                    Event.new(EventAction.CONFIGURATION_ERROR, message=message).save()
```

PKizzle replied:
Fixed.

BeryJu committed:
Update authentik/lib/config.py

Signed-off-by: Jens L. <jens@beryju.org>

PKizzle committed:
Fix Event deprecation handling

PKizzle committed:
Merge branch 'feature/reorganize-redis-config' of github.com:PKizzle/authentik into feature/reorganize-redis-config

PKizzle commented:
@BeryJu Are there any other changes you would like to see?
